mlx/io/gguf.cpp

```diff
@@ -70,6 +79,72 @@ std::tuple<allocator::Buffer, Dtype> extract_tensor_data(gguf_tensor* tensor) {
   return {buffer, float16};
 }
 
+void extract_q4_1_data(
+    std::unordered_map<std::string, array>* a,
+    const gguf_tensor& tensor) {
+  std::string name = std::string(tensor.name, tensor.namelen);
+  const std::vector<int> shape = get_shape(tensor);
+  const uint64_t weights_per_byte = 2;
+  const uint64_t weights_per_block = 32;
+  if (shape[shape.size() - 1] % weights_per_block != 0) {
+    std::ostringstream msg;
+    msg << "[save_gguf] tensor " << name
+        << "has incompatible last dim shape: " << shape[shape.size() - 1];
+    throw std::runtime_error(msg.str());
+  }
+  const uint64_t bytes_per_block =
+      20; // 2 bytes scale, 2 bytes bias, 32x0.5 byte weights
+  const uint64_t num_blocks = tensor.num_weights / weights_per_block;
+  allocator::Buffer weights_buffer =
+      allocator::malloc(tensor.num_weights / weights_per_byte);
+  allocator::Buffer scales_buffer = allocator::malloc(num_blocks * 2);
+  allocator::Buffer biases_buffer = allocator::malloc(num_blocks * 2);
+  auto data = (uint8_t*)tensor.weights_data;
+  auto weigths = (uint8_t*)weights_buffer.raw_ptr();
+  auto scales = (uint16_t*)scales_buffer.raw_ptr();
+  auto biases = (uint16_t*)biases_buffer.raw_ptr();
+  for (int64_t i = 0; i < num_blocks; i++) {
+    uint8_t* block_data = data + i * bytes_per_block;
+    scales[i] = *((uint16_t*)block_data);
+    biases[i] = *((uint16_t*)block_data + 1);
+    // 16 weights in lower bits
+    for (int64_t j = 0; j < 16; ++j) {
+      uint8_t x = (block_data[j + 4] & 0x0F);
+      if (j % 2 == 0) {
+        x <<= 4;
+      }
+      weigths[i * 16 + j / 2] += x;
+    }
+    // 16 weights in higher bits
+    for (int64_t j = 0; j < 16; ++j) {
+      uint8_t x = (block_data[j + 4] >> 4);
+      if (j % 2 == 0) {
+        x <<= 4;
+      }
+      weigths[i * 16 + 8 + j / 2] += x;
+    }
+  }
+  std::vector<int> weights_shape = shape;
+  weights_shape[shape.size() - 1] =
+      weights_shape[shape.size() - 1] / weights_per_byte / 4;
+  a->insert({name, array(weights_buffer, weights_shape, uint32)});
+
+  const std::string weight_suffix = ".weight";
+  const std::string name_prefix =
+      name.substr(0, name.length() - weight_suffix.length());
+  std::vector<int> scale_bias_shape = shape;
+  scale_bias_shape[shape.size() - 1] =
+      scale_bias_shape[shape.size() - 1] / weights_per_block;
+
+  const std::string scales_name =
+      (std::ostringstream() << name_prefix << ".scales").str();
+  a->insert({scales_name, array(scales_buffer, scale_bias_shape, float16)});
+
+  const std::string biases_name =
+      (std::ostringstream() << name_prefix << ".biases").str();
+  a->insert({biases_name, array(biases_buffer, scale_bias_shape, float16)});
```

Comment:
I picked these names to match the names of QuantizedLinear modules, so that the model can be initialized without additional remapping.

Name collisions are possible, but unlikely. GGUF supports a small set of architectures and the convention seems to be that all FF weights have this `.weight` suffix. We could add some assertions to raise an error on collisions though.

Reply:
Understood, but this layer isn't really supposed to know anything about `QuantizedLinear` . It couples things in a way that makes it slightly harder to change. For example what if I want to serialize a quantized matrix that is called `myfavoritename`? It would not be possible to deserialize it with this approach.

Reply:
Fair point :)